### DISCUSSION TITLE
Mass payouts contract

### DISCUSSION THREAD

**Discussion by earrietadev** [stellar/sorobanathon#114] [General]
> First of all: this is both my first Soroban contract created and the first time I do something with Rust and so far everything has been easier than expected so I feel that's good for future adoption... That being said let's get to the topic:

Every month I need to send funds to multiple accounts in order to pay for businesses expenses, having a contract I can set in a way it sends to all recipients automatically every time I triggers it seemed like a good first contract to do.

Contract can be found at https://github.com/earrietadev/soroban-mass-payouts-contract/ but here is a quick view of what it can do:

```rust
pub trait MassPayoutContractTrait {

  // This method is what it sets the contract configuration
  // Most methods in the contract won't work if this method hasn't been called before
  // Method can only be called once
  // "admin" can be other than the deployer of the contract
  // "currency" is the token contract id which you will pay recipients with
  fn initialize(
    env: Env,
    admin: Address,
    currency: BytesN<32>,
  );

  // This method is used to check if the method "initialize" was already called
  // If not initialized, the method will throw an error
  fn init_done(env: Env);

  // A method to return the current state of the contract
  // This method is meant to be used internally by the contract
  fn get_state(env: Env) -> State;

  // This method returns the amount an account is set to receive during a payout
  fn get_acc(env: Env, account: AccountId) -> u32;

  // The amount a recipient will receive during a payout is done with this method
  // It can be called multiple times for the same account, the value will be updated
  // This method requires the contract had been initialized already
  // This method can only be called by the admin of the contract
  fn set_acc(env: Env, account: AccountId, amount: u32);

  // With this method the invoker will deposit funds into the contract
  // Depositor needs to approve with the currency contract before calling this method
  // This method requires the contract had been initialized already
  fn deposit(env: Env, amount: u32);

  // This method starts the payout process to all accounts set in the contract
  // If there is not enough funds in the contract then it will throw an error
  // This method requires the contract had been initialized already
  // This method can only be called by the admin of the contract
  fn payout(env: Env);
}
```

I also included a few scripts in the `Makefile` and a quick guide in the `README` which explains how to interact with the contract or deploying it to the network.

Any feedback is welcome!

**Comment by achallagundla** [comment]
Heya,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 1100 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya

PS: No need to submit the form if you already have, just update me on discord if your wallet address has changed.
